core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,465 @@
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link LoadBalancer} that provides no load-balancing over the addresses from the {@link
+ * io.grpc.NameResolver}.  The channel's default behavior is used, which is walking down the address
+ * list and sticking to the first that works.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/10383")
+final class PickFirstLeafLoadBalancer extends LoadBalancer {
+  private final Helper helper;
+  private final Map<SocketAddress, Subchannel> subchannels = new HashMap<>();
+  private final Map<Subchannel, ConnectivityState> states = new HashMap<>();
+  private Index addressIndex;
+  private volatile ConnectivityState currentState = IDLE;
+
+  PickFirstLeafLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+  }
+
+  @Override
+  public boolean acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    List<EquivalentAddressGroup> servers = resolvedAddresses.getAddresses();
+    if (servers.isEmpty()) {
+      handleNameResolutionError(Status.UNAVAILABLE.withDescription(
+          "NameResolver returned no usable address. addrs=" + resolvedAddresses.getAddresses()
+          + ", attrs=" + resolvedAddresses.getAttributes()));
+      return false;
+    }
+    for (EquivalentAddressGroup eag : servers) {
+      if (eag == null) {
+        handleNameResolutionError(Status.UNAVAILABLE.withDescription(
+            "NameResolver returned address list with null endpoint. addrs="
+            + resolvedAddresses.getAddresses() + ", attrs=" + resolvedAddresses.getAttributes()));
+        return false;
+      }
+    }
+    // We can optionally be configured to shuffle the address list. This can help better distribute
+    // the load.
+    if (resolvedAddresses.getLoadBalancingPolicyConfig()
+        instanceof PickFirstLeafLoadBalancerConfig) {
+      PickFirstLeafLoadBalancerConfig config
+          = (PickFirstLeafLoadBalancerConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+      if (config.shuffleAddressList != null && config.shuffleAddressList) {
+        servers = new ArrayList<EquivalentAddressGroup>(servers);
+        Collections.shuffle(servers,
+            config.randomSeed != null ? new Random(config.randomSeed) : new Random());
+      }
+    }
+
+    final List<EquivalentAddressGroup> newImmutableAddressGroups =
+        Collections.unmodifiableList(new ArrayList<>(servers));
+
+    if (addressIndex == null) {
+      addressIndex = new Index(newImmutableAddressGroups);
+    // If a ready connection exists in new address list,
+    // keep this connection and don't create new subchannels
+    } else {
+      SocketAddress previousAddress = addressIndex.getCurrentAddress();
+      addressIndex.updateGroups(newImmutableAddressGroups);
+      if (addressIndex.seekTo(previousAddress)
+          && states.get(subchannels.get(previousAddress)) == READY) {
```

Comment:
We need to compare to the previousAddress in case the LB is ready but on a different subchannel. I don't think it's necessarily true that the current address is always on the READY subchannel in case another subchannel comes out of backoff and readies while we are iterating.

Also, `subchannels.get(previousAddress)` should not be null anymore after the change we discussed about removing the hackiness of that.

Reply:
It is true that when it turns to READY, the addressIndex is not necessarily pointing to that address. After we removed the resetting index, it can be at the invalid position. So, L100 would no longer work.

What about always putting index to the ready subchannel when READY, then can we simplify to this? 
```
if (currentState == READY && addressIndex.seekTo(previousAddress)) {
  return true;
}
``` 
Also the L109 addressIndex.reset() seems can be removed.

Alternatively we can keep resetting the index it goes READY. And we can have a `activeSubchannel` saved so we can compare if the previous ready channel exists in the current list.

Reply:
What is the desired behavior in the case where we are READY and a connection breaks? Currently, I see that when the `transportShutdown()` method is called, we go into an IDLE state if we have an active transport. 

I think that's likely the only edge case to worry about?